packages/legend-graph/src/helpers/DomainHelper.ts

```diff
@@ -46,6 +58,91 @@ export const _package_deleteElement = (
   );
 };
 
+export const getElementRootPackage = (element: PackageableElement): Package =>
+  !element.package
+    ? guaranteeType(element, Package)
+    : getElementRootPackage(element.package);
+
+/**
+ * If package name is a path, continue to recursively
+ * traverse the package chain to find the leaf package
+ *
+ * NOTE: if we do not allow create new packages, errorcould be
+ * thrown if a package with the specified path is not found
+ */
+export const getOrCreatePackage = (
+  parentPackage: Package,
+  relativePackagePath: string,
+  createNewPackageIfNotFound: boolean,
+  cache: Map<string, Package> | undefined,
+): Package => {
+  const index = relativePackagePath.indexOf(ELEMENT_PATH_DELIMITER);
+  const packageName =
+    index === -1
+      ? relativePackagePath
+      : relativePackagePath.substring(0, index);
+  const packagePath = createPath(parentPackage.fullPath, packageName);
+
+  // check cache to find the package
+  // NOTE: we only return from cache if there is no child package to further traverse
```

Comment:
because we're traversing the chain of package in order to get (or create and then get) the package:

`a::b::c`, we're traversing `a` then `a::b` then `a::b::c`, here I just short-circuit `a::b::c`, but maybe I can also do some more useful lookup to improve the speed.

Reply:
yeah, this looks dumb though, now that I think about the `packagePath === -1`, doesn;t do much.

Reply:
addressed. nice catch! I have improved the algo now